aten/src/ATen/ParallelNativeMobile.h

```diff
@@ -0,0 +1,122 @@
+#pragma once
+
+#include <algorithm>
+#include <cstddef>
+#include <exception>
+
+#define INTRA_OP_PARALLEL
+
+namespace at {
+namespace internal {
+
+// Run lambda function `fn` over `task_id` in [0, `range`) with threadpool.
+// `fn` will be called with params: (thread_pool_task_id, task_id).
+CAFFE2_API void _run_with_pool(const std::function<void(int, size_t)>& fn, size_t range);
+
+// RAII guard helps to support in_parallel_region() and get_thread_num() API.
+struct CAFFE2_API ParallelRegionGuard {
+  ParallelRegionGuard(int64_t task_id);
+  ~ParallelRegionGuard();
+};
+
+inline std::tuple<size_t, size_t> calc_num_tasks_and_chunk_size(
+    int64_t begin, int64_t end, int64_t grain_size) {
+  if ((end - begin) < grain_size) {
+    return std::make_tuple(1, std::max((int64_t)0, end - begin));
+  }
+  // Choose number of tasks based on grain size and number of threads.
+  size_t chunk_size = divup((end - begin), get_num_threads());
+  // Make sure each task is at least grain_size size.
+  chunk_size = std::max((size_t)grain_size, chunk_size);
+  size_t num_tasks = divup((end - begin), chunk_size);
+  return std::make_tuple(num_tasks, chunk_size);
+}
+
+} // namespace internal
+
+template <class F>
+inline void parallel_for(
+    const int64_t begin,
+    const int64_t end,
+    const int64_t grain_size,
+    const F& f) {
+  TORCH_CHECK(grain_size >= 0);
+  if (begin >= end) {
+    return;
+  }
+
+  if (((end - begin) >= grain_size) && !in_parallel_region()) {
+    size_t num_tasks, chunk_size;
+    std::tie(num_tasks, chunk_size) =
+        internal::calc_num_tasks_and_chunk_size(begin, end, grain_size);
+    std::atomic_flag err_flag = ATOMIC_FLAG_INIT;
+    std::exception_ptr eptr;
+    std::vector<c10::ivalue::Future> futures(num_tasks);
+    auto task = [f, &eptr, &err_flag, &futures, begin, end, chunk_size]
+        (int idx, size_t task_id) {
+      int64_t local_start = begin + task_id * chunk_size;
+      if (local_start < end) {
+        int64_t local_end = std::min(end, (int64_t)(chunk_size + local_start));
+        try {
+          internal::ParallelRegionGuard guard(task_id);
+          f(local_start, local_end);
+        } catch (...) {
+          if (!err_flag.test_and_set()) {
+            eptr = std::current_exception();
+          }
+        }
+      }
+      futures[task_id].markCompleted();
+    };
+    internal::_run_with_pool(task, num_tasks);
```

Comment:
we don't utilize the main thread to do some compute, do we?

Reply:
caffe2's thread pool (gemmlowp clone) runs one task (first one) on current thread directly: https://github.com/pytorch/pytorch/blob/master/caffe2/utils/threadpool/WorkersPool.h#L345

That's one of the reasons I decided to fork and modify ParallelMobile.h (which also runs on main thread explicitly) - doing it at both places is wrong.